sdb/commands/container_of.py

```diff
@@ -0,0 +1,65 @@
+#
+# Copyright 2020 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+#
+# The pylint workaround below is for the example
+# sections of the help messages.
+#
+# pylint: disable=line-too-long
+
+import argparse
+from typing import Iterable
+
+import drgn
+import sdb
+from sdb.commands.internal.util import get_valid_struct_name
+
+
+class ContainerOf(sdb.Command):
+    """
+    Get the containing object of a pointer object.
+
+    EXAMPLES
+        Trivial example of getting a structure's address
+        out of the address of one of its members:
+
+            sdb> addr init_task | cast void *
+            (void *)0xffffffffa8217740
+            sdb> addr init_task | member comm | addr | container_of struct task_struct comm  | cast void *
```

Comment:
I don't mean for this to be a blocking comment.. but I'm curious.. do we always have to provide the structure and the member?

e.g. if we did this instead:
```
addr init_task | member comm | container_of  | cast void *
```
could this work?

I'm wondering, since we probably could (not sure if we actually do) have all the type information passed across the pipes..

e.g. `addr init_task` should know the type of `init_task` is `struct task_struct`.. and then `member` knows it's pulling a member from a `struct task_struct`.. so then, perhaps `container_of` could know that it was passed the `comm` member of the `struct task_struct` structure, just from the information in the pipe, and automatically do the right thing..

I don't know if all of the necessary information is passed into `container_of` right now, but I thought I'd at least ask.. and if not, would it be useful to try and achieve something like this (i.e. open a bug, and leave it for later work)?

Reply:
`container_of` only get the object that's passed to it, which doesn't know what its "container" is.  We could theoretically make `member` pass this information through somehow, but this is a contrived use.  If you have the container already you wouldn't use `|member ... | container_of`.  Typically you would have the struct pointer from some other source and happen to know that it's inside of another struct.  E.g. I have an `avl_node_t*` which I happen to know is `dmu_buf_impl_t`'s `db_link` and I want to get the containing `dmu_buf_impl_t*`.

So I don't think it's practical for `container_of` to figure out what the containing type is.

Reply:
:+1: yea, ok.. I think that makes sense.